### PR TITLE
Add functionality to deprecate argument rather than rename

### PR DESCRIPTION
**Pull request by bsipocz**
We currently have no way to deprecate an argument. I think it's preferable to reuse what we have, but I'm happy to add a new `deprecate_argument` decorator if people prefer that instead of this hack.

As for other utilities useful for downstream, I prefer to milestone it for LTS.

### CODE REVIEW

**Comment by bsipocz** [comment]
I'll add tests and changelog once it gets the generic thumps up.

File: astropy/utils/decorators.py [hunk 1]
```diff
@@ -275,7 +275,8 @@ def delete(self):
 def deprecated_renamed_argument(old_name, new_name, since,
                                 arg_in_kwargs=False, relax=False,
                                 pending=False,
-                                warning_type=AstropyDeprecationWarning):
+                                warning_type=AstropyDeprecationWarning,
+                                alternative=''):
     """Deprecate a _renamed_ function argument.
```
Comment (mhvk) [inline astropy/utils/decorators.py]: `_renamed_ or _removed_`

File: astropy/utils/decorators.py [hunk 2]
```diff
@@ -318,6 +320,11 @@ def deprecated_renamed_argument(old_name, new_name, since,
         Warning to be issued.
         Default is `~astropy.utils.exceptions.AstropyDeprecationWarning`.
 
+    alternative : str, optional
+        An alternative function or class name that the user may use in
+        place of the deprecated object if ``new_name`` is None. The deprecation
```
Comment (mhvk) [inline astropy/utils/decorators.py]: I think I'd also use this alternative if `new_name` is given, i.e., make this something that overrides `new_name` in the message.

File: astropy/utils/decorators.py [hunk 3]
```diff
@@ -467,15 +477,19 @@ def wrapper(*args, **kwargs):
...
-                            'Use argument "{2}" instead.'
-                            ''.format(old_name[i], since[i], new_name[i]),
-                            warning_type, stacklevel=2)
+                        message = ('"{0}" was deprecated in version {1} '
+                                   'and will be removed in a future version. '
+                                   .format(old_name[i], since[i]))
+                        if new_name[i] is not None:
+                            message += ('Use argument "{}" instead.'
+                                        .format(new_name[i]))
+                        elif alternative:
```
Comment (mhvk) [inline astropy/utils/decorators.py]: So, if we go with my suggestion, I'd swap the two branches of this `if` statement.

**Comment by pllim** [comment]
Pinging @MSeifert04 since he added deprecation functionality a while back (in my vague memory).

File: astropy/utils/decorators.py [hunk 4]
```diff
@@ -496,6 +510,8 @@ def wrapper(*args, **kwargs):
                                 raise TypeError(
                                     'cannot specify both "{}" and "{}"'
                                     '.'.format(old_name[i], new_name[i]))
+                    elif new_name[i] is None:
```
Comment (MSeifert04) [inline astropy/utils/decorators.py]: I would combine this with the `else` or move the comment (the "# If the new argument isn't specified..."), otherwise this will be a bit hard to follow.

**Comment by MSeifert04** [comment]
> since he added deprecation functionality a while back (in my vague memory).

Yeah, I've added that function, but I didn't remember it being that complex. 😓

File: astropy/utils/decorators.py [hunk 5]
```diff
@@ -467,15 +477,19 @@ def wrapper(*args, **kwargs):
                 # parameter was renamed to newkeyword.
                 if old_name[i] in kwargs:
                     value = kwargs.pop(old_name[i])
-                    # Display the deprecation warning only when it's only
+                    # Display the deprecation warning only when it's not
```
Comment (MSeifert04) [inline astropy/utils/decorators.py]: Good catch 👍

**Comment by bsipocz** [comment]
> Yeah, I've added that function, but I didn't remember it being that complex.

I share the feeling, and it even has a name :) "Eagleson's law".

Anyway, the question is that you are OK with making it a bit more complex rather than creating a new decorator instead?

**Comment by MSeifert04** [comment]
> Anyway, the question is that you are OK with making it a bit more complex rather than creating a new decorator instead?

It's actually not adding much complexity, however I fear that the combination of alternative and new_name might be tricky (depending on how we decide how it should work). I probably should've written a few helper functions instead of putting it all in the function itself to reduce the complexity (or maybe I had a reason not to ...).

File: astropy/utils/decorators.py [hunk 6]
```diff
@@ -289,8 +290,9 @@ def deprecated_renamed_argument(old_name, new_name, since,
     old_name : str or list/tuple thereof
         The old name of the argument.
 
-    new_name : str or list/tuple thereof
-        The new name of the argument.
+    new_name : str or list/tuple thereof or `None`
+        The new name of the argument. ``old_name`` has been deprecated to
+        be removed if `None`.
```
Comment (astrofrog) [inline astropy/utils/decorators.py]: Maybe rephrase to ``The new name of the argument. Set this to `None` if removing the argument instead of renaming.``

File: astropy/utils/decorators.py (hunk 2 above)
Comment (bsipocz) [reply astropy/utils/decorators.py]: I think I'm with @MSeifert04 here, it's either to use the new name or why bother renaming rather than removing if there is a preferred alternative elsewhere?

**Comment by bsipocz** [comment]
I would appreciate a final review now that a test is added and the logic is cleaned up based on the review and the test itself.

File: astropy/utils/decorators.py (hunk 3 above)
Comment (pllim) [reply astropy/utils/decorators.py]: Was this addressed already?

Comment (bsipocz) [reply astropy/utils/decorators.py]: No, and currently I don't plan to as per the discussion above https://github.com/astropy/astropy/pull/8324#discussion_r258758232. I happy to though in case there is a clear use case comes up above.

**Comment by bsipocz** [comment]
The linkchecker sadly double runs the docs building, but this makes it reactively easy to see that the issue is indeed with the links, as the docs build step green in the workflow above it:

screen shot 2019-02-21 at 12 11 43

**Comment by pllim** [comment]
I am merging this then. Thanks!

**Comment by bsipocz** [comment]
Thanks @pllim!

**Comment by bsipocz** [comment]
@mhvk - In case you have a use case in mind for the question above (https://github.com/astropy/astropy/pull/8324#discussion_r245425763), I'm happy to address it in a follow-up PR before getting this release out.:

>You mean `new_name` as the new argument name but providing a different name as alternative? That would be odd, do you have a use-case in mind where this would be helpful? I'd say it would be more confusing to provide a different name.
